client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -252,33 +254,46 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = {
-          val entity = req.entity match {
-            case Entity.Empty | Entity.Strict(_) =>
-              req
-            case Entity.Default(_, _) =>
-              req.pipeBodyThrough(until(disposed))
-          }
-
-          addHostHeaderIfUriIsAbsolute(entity)
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message.entity match {
+      case Entity.Empty | Entity.Strict(_) =>
+        message match {
+          case req: Request[F] =>
+            val reqAugmented = addHostHeaderIfUriIsAbsolute(req)
+            Resource.eval(app(reqAugmented)).flatMap(run(_))
+          case resp: Response[F] =>
+            Resource.eval(F.pure(resp))
         }
-
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map { resp =>
-            resp.entity match {
-              case Entity.Empty | Entity.Strict(_) =>
-                resp
-              case Entity.Default(_, _) =>
-                resp.pipeBodyThrough(until(disposed))
+      case Entity.Default(_, _) =>
+        message match {
+          case req: Request[F] =>
+            Resource.suspend(
+              Ref[F].of(false).map { disposed =>
+                val reqAugmented =
+                  addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+                Resource
+                  .eval(app(reqAugmented))
+                  .onFinalize(disposed.set(true))
+                  .flatMap(run(_, Option(disposed)))
+              }
+            )
+          case resp: Response[F] =>
+            refOp.fold {
+              Resource.suspend(Ref[F].of(false).map { disposed =>
+                Resource
+                  .eval(F.pure(resp.pipeBodyThrough(until(disposed))))
+                  .onFinalize(disposed.set(true))
+              })
+            } { disp =>
+              Resource
+                .eval(F.pure(resp.pipeBodyThrough(until(disp))))
+                .onFinalize(disp.set(true))
```

Comment:
I'm a little confused by these lines.
1. `Resource.eval(F.pure(...))` is equivalent to `Resource.pure(...)`
2. Since the code is the same, and there is just the issue of whether the `Ref` is provided or not, I think you can do `refOp.fold(Ref[F].of(false))(F.pure).flatMap { disp => ... }` and then share the code.

Reply:
@armanbilge I think I fixed the duplication, let me know if you'd like additional changes.  Thanks for the review!